File: MyNotesProject/MyNotes/src/main/java/com/fsk/mynotes/services/SaveNoteService.java
package com.fsk.mynotes.services;


import android.app.IntentService;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;

import com.fsk.common.database.DatabaseHelper;
import com.fsk.common.utils.Preconditions;
import com.fsk.mynotes.constants.NoteExtraKeys;
import com.fsk.mynotes.data.Note;


/**
 * A service that saves a note to persistent storage.
 */
public class SaveNoteService extends IntentService {

    /**
     * Start the service.
     *
     * @param context The context to use for starting the service.
     * @param note The note to save.
     */
    public static void startService(@NonNull Context context, @NonNull Note note) {
        Preconditions.checkNotNull(context);
        Preconditions.checkNotNull(note);

        Intent intent = new Intent(context, SaveNoteService.class);
        intent.putExtra(NoteExtraKeys.NOTE_KEY, note);
        context.startService(intent);
    }


    /**
     * Default Constructor.
     */
    public SaveNoteService() {
        super(SaveNoteService.class.getName());
    }


    @Override
    protected void onHandleIntent(final Intent intent) {
        Note note = intent.getParcelableExtra(NoteExtraKeys.NOTE_KEY);
        note.save(DatabaseHelper.getDatabase());
    }
}

File: MyNotesProject/MyNotes/src/main/java/com/fsk/mynotes/services/SaveNoteService.kt
package com.fsk.mynotes.services;


import android.app.IntentService
import android.content.Context
import android.content.Intent
import com.fsk.mynotes.constants.NOTE_KEY_EXTRA
import com.fsk.mynotes.data.Note


/**
 * Start the service.
 *
 * @param note The note to save.
 */
fun Context.startSaveNoteService(note: Note) {
    val intent = Intent(this, SaveNoteService::class.java).apply {
        putExtra(NOTE_KEY_EXTRA, note)
    }
    startService(intent);
}


/**
 * A service that saves a note to persistent storage.
 */
class SaveNoteService() : IntentService(SaveNoteService::class.java.simpleName) {


    override fun onHandleIntent(intent: Intent?) {
        if (intent != null) {
            val note: Note = intent.getParcelableExtra(NOTE_KEY_EXTRA);
            if (note.text?.isNotEmpty()) {
                note.save(contentResolver);
            }
            else {
                note.delete(contentResolver);
            }
        }
    }


}
